wasm/src/main/scala/org/scalajs/linker/backend/webassembly/BinaryWriter.scala

```diff
@@ -2,53 +2,60 @@ package org.scalajs.linker.backend.webassembly
 
 import scala.annotation.tailrec
 
-import org.scalajs.ir.Position
+import org.scalajs.ir.{Position, UTF8String}
 import org.scalajs.linker.backend.javascript.SourceMapWriter
 
 import Instructions._
-import Names._
+import Identitities._
 import Modules._
 import Types._
 
 class BinaryWriter(module: Module, emitDebugInfo: Boolean) {
   import BinaryWriter._
 
-  private val typeIdxValues: Map[TypeName, Int] =
-    module.types.flatMap(_.subTypes).map(_.name).zipWithIndex.toMap
+  private val typeIdxValues: Map[TypeID, Int] =
+    module.types.flatMap(_.subTypes).map(_.id).zipWithIndex.toMap
 
-  private val dataIdxValues: Map[DataName, Int] =
-    module.datas.map(_.name).zipWithIndex.toMap
+  private val dataIdxValues: Map[DataID, Int] =
+    module.datas.map(_.id).zipWithIndex.toMap
 
-  private val allFunctionNames: List[FunctionName] = {
-    val importedFunctionNames = module.imports.collect {
-      case Import(_, _, ImportDesc.Func(id, _)) => id
+  private val funcIdxValues: Map[FunctionID, Int] = {
+    val importedFunctionIDs = module.imports.collect {
+      case Import(_, _, ImportDesc.Func(id, _, _)) => id
     }
-    importedFunctionNames ::: module.funcs.map(_.name)
+    val allIDs = importedFunctionIDs ::: module.funcs.map(_.id)
+    allIDs.zipWithIndex.toMap
   }
 
-  private val funcIdxValues: Map[FunctionName, Int] =
-    allFunctionNames.zipWithIndex.toMap
-
-  private val tagIdxValues: Map[TagName, Int] = {
-    val importedTagNames = module.imports.collect { case Import(_, _, ImportDesc.Tag(id, _)) =>
+  private val tagIdxValues: Map[TagID, Int] = {
+    val importedTagIDs = module.imports.collect { case Import(_, _, ImportDesc.Tag(id, _, _)) =>
       id
     }
-    val allNames = importedTagNames ::: module.tags.map(_.name)
-    allNames.zipWithIndex.toMap
+    val allIDs = importedTagIDs ::: module.tags.map(_.id)
+    allIDs.zipWithIndex.toMap
   }
 
-  private val globalIdxValues: Map[GlobalName, Int] = {
-    val importedGlobalNames = module.imports.collect {
-      case Import(_, _, ImportDesc.Global(id, _, _)) => id
+  private val globalIdxValues: Map[GlobalID, Int] = {
+    val importedGlobalIDs = module.imports.collect {
+      case Import(_, _, ImportDesc.Global(id, _, _, _)) => id
     }
-    val allNames = importedGlobalNames ::: module.globals.map(_.name)
-    allNames.zipWithIndex.toMap
+    val allIDs = importedGlobalIDs ::: module.globals.map(_.id)
+    allIDs.zipWithIndex.toMap
+  }
+
+  private val fieldIdxValues: Map[TypeID, Map[FieldID, Int]] = {
+    (for {
+      recType <- module.types
+      SubType(typeID, _, _, _, StructType(fields)) <- recType.subTypes
+    } yield {
+      typeID -> fields.map(_.id).zipWithIndex.toMap
+    }).toMap
```

Comment:
Great, I like having name resolution for field access only in the binary writer (instead of `getFieldIdx` everywhere) looks much readable!